### PR TITLE
cache split footer

### LINKED ISSUE #444: Cache the bundle footer
We used to have a mechanism to cache the hotcache.

It will be lost now that we moved to the bundle system.
We want a cache to be able to cache the entire footer.

### PR DESCRIPTION
**Pull request by PSeitz**
fix quickwit-search tests
closes #444

### CODE REVIEW

File: quickwit-search/src/leaf.rs [hunk 1]
```diff
@@ -21,28 +21,38 @@
...
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
 use tantivy::{collector::Collector, query::Query, Index, ReloadPolicy, Searcher, Term};
 use tokio::task::spawn_blocking;
 
-/// Opens a `tantivy::Index` for the given split.
-///
-/// The resulting index uses a dynamic and a static cache.
-pub(crate) async fn open_index(
+lazy_static! {
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-search/src/leaf.rs:39]: once_cell>lazy_static.

No macro involved.

File: quickwit-search/src/leaf.rs [hunk 2]
```diff
@@ -21,28 +21,38 @@
...
+    static ref SPLIT_FOOTER_CACHE: Mutex<LruCache<String, Bytes>> = Mutex::new(LruCache::new(500));
```
Comment (fulmicoton) [inline quickwit-search/src/leaf.rs:40]: We have code to have an LruCache for which the limit is expressed in bytes. It is clunky, but can we use that?

File: quickwit-search/src/leaf.rs [hunk 3]
```diff
@@ -56,6 +66,25 @@ pub(crate) async fn open_index(
...
+/// Opens a `tantivy::Index` for the given split.
+///
+/// The resulting index uses a dynamic and a static cache.
+pub(crate) async fn open_index(
+    index_storage: Arc<dyn Storage>,
+    split_and_footer_offsets: &SplitIdAndFooterOffsets,
+) -> anyhow::Result<Index> {
+    let split_file = PathBuf::from(format!("{}.split", split_and_footer_offsets.split_id));
+    let mut footer_data =
+        get_or_load_split_footer(index_storage.clone(), split_and_footer_offsets).await?;
```
Comment (fulmicoton) [inline quickwit-search/src/leaf.rs:87]: `get_from_cache_or_fetch`

File: quickwit-search/src/leaf.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit-search/src/leaf.rs:39]: `lazy_static` is more readable, what's the problem with a macro in that case?

File: quickwit-search/src/leaf.rs [hunk 4]
```diff
@@ -21,28 +21,38 @@
...
+}
+
+async fn get_or_load_split_footer(
     index_storage: Arc<dyn Storage>,
     split_and_footer_offsets: &SplitIdAndFooterOffsets,
-) -> anyhow::Result<Index> {
+) -> anyhow::Result<Bytes> {
+    {
+        let mut cache_mutex_guard = SPLIT_FOOTER_CACHE.lock().unwrap();
+        let possible_val = cache_mutex_guard.get(&split_and_footer_offsets.split_id);
```
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit-search/src/leaf.rs:49]: yep

File: quickwit-search/src/leaf.rs (hunk 2 above)
Comment (PSeitz) [reply quickwit-search/src/leaf.rs:40]: I need to make it more generic first, but yes

File: quickwit-search/src/leaf.rs (hunk 3 above)
Comment (PSeitz) [reply quickwit-search/src/leaf.rs:87]: yep

File: quickwit-search/src/leaf.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PSeitz) [reply quickwit-search/src/leaf.rs]: I changed it to once_cell

File: quickwit-search/src/leaf.rs [hunk 5]
```diff
@@ -21,28 +21,39 @@
...
 use tantivy::{collector::Collector, query::Query, Index, ReloadPolicy, Searcher, Term};
 use tokio::task::spawn_blocking;
 
-/// Opens a `tantivy::Index` for the given split.
-///
-/// The resulting index uses a dynamic and a static cache.
-pub(crate) async fn open_index(
+fn global_split_footer_cache() -> &'static MemorySizedCache<String> {
+    static INSTANCE: OnceCell<MemorySizedCache<String>> = OnceCell::new();
+    INSTANCE.get_or_init(|| MemorySizedCache::with_capacity_in_bytes(25_000_000))
```
Comment (fulmicoton) [inline quickwit-search/src/leaf.rs:42]: 25 MB is way too small. Can you make it 1GB?

File: quickwit-search/src/leaf.rs [hunk 6]
```diff
@@ -21,28 +21,39 @@
...
+}
+
+async fn get_from_cache_or_fetch(
```
Comment (fulmicoton) [inline quickwit-search/src/leaf.rs:45]: `get_split_footer_from_cache_or_fetch`

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**